Installation and upgrade worflows for RDM

ppanero opened the pull request.

lnielsen commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,100 @@
+- As a developer, I want to be able to upgrade my running instance to a newer versions of Invenio RDM.
+
+## Detailed design
+
```
I would add a section about how it would look like for an end-user

```
$ invenio init
... <how should the output/questions look like> ...
$ invenio build
...
$ invenio run
```

lnielsen replied:
Also, how do they get the tool? (``pip install ...`` ?) And what are the requirements (Docker, Python 3, COmpose?)

lnielsen commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,100 @@
+- `run`
+- `update`
+- `upgrade`
+- `destroy`
```
I think we also need a ``setup`` command? Something a la :

```
$ invenio setup
Creating queues on message broker...
Creating database...
Creating indexes...
Loading initial users, roles and permissions
```

lnielsen commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,100 @@
+
+## Detailed design
+
+### CLI and commands
```
Important question - what do we name it :-)

fenekku commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,160 @@
+- Is the division between base and application image needed? Does docker build caching system solve the problem we are trying to address?
+- Command examples might vary in the future (e.g. CLI name might change). Maybe is better to be put in the module documentation. To allow deprecation, etc.
+- Shall we create an RFC for CLI and then a specific RFC for each flavour that can be installed with the corresponding CLI?
+- "semi-production" is used along the RFC, shall it be changed by "pseudo". The word production might give the impresion that it can be run as is. However, there are many other things to take into account in a production deployment (replication, persistency, HA of each component, security, etc.)
```
Yes, these were also all the questions I had.

ppanero commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,160 @@
+
+**init**
+
+The `init` command would initialize the files and folders structure, resulting in a working instance. The initialization will be performed by prompting the user with a set of questions that allow to customize the system architecture (i.e. project name, Elasticsearch version, relational database system). This boostraping is implemented as a Cookiecutter module.
```
Yes, totally correct. Should be e.g.

ppanero commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,160 @@
+
+- `invenio.cfg`: Contains all the configuration variables.
+- `pipenv`: Contains the python dependencies along with each corresponding version.
+- `static/`: Contains all the CSS and potential static content (e.g. images).
```
I am inclined to say yes, but I do not want to include it in the RFC until the decision has been taken (on which path to go). So probably we can amend this RFC in January for this part? What do you think?

fenekku replied:
Yes, that seems reasonable to omit it for now and wait until we are actually working on the Javascript inclusion.

ppanero commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,160 @@
+
+**build**
+
+The `build` command should allow to build two types of environments: Development and semi-production.
```
Agree, thanks!

ppanero commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,160 @@
+
+The `build` command should allow to build two types of environments: Development and semi-production.
+
+Nonetheless, independently from the environment type, the first step of the build process is to lock the application dependencies. Therefore, by using Pipenv, the CLI will generate a *lock* file that would contain the version and hash identifier of each dependency. For this use case the flags ``--lock`` or ``--no-lock`` are used, being the first the default one.
```
Yes... :( The problem is also that the lock happens outside of docker build. Since the lock file has to be copied over to the container (so it can be tracked via VCS). So we must allow for ``--skip-lock``.

ppanero commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,160 @@
+
+**upgrade**
+
+The `upgrade` command allows users to upgrade there instances, in terms of Invenio RDM versions. This will require the rebuild of the application image, and potentially the base image. Since this might also carry mayor canges, e.g. in the datamodel, the Elasticsearch and Database services would need to be updated.
```
You are right. What about "significant"? In order not to mislead the user.

ppanero commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,160 @@
+
+## Motivation
+
+- As a developer, I want to be able to initialize the RDM file/folder structure, so that I can have a working instance.
```
I agree. I had it just for consistency with other RFCs. Lets discuss IRL

ppanero commented:
@fenekku Thanks for the review! I fixed/reply to the comments. I also added one unnresolved question that could potentially affect your profiles/falvours RFC.

fenekku commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,161 @@
+
+It has a series of hard dependencies that should be installed by the user:
+
+- Python 3.4+
```
Now 3.5+, right? (Given that 3.4 has [reached end of life](https://www.python.org/downloads/release/python-340/))

ppanero replied:
:+1: indeed, thanks for the catch

ppanero commented on rdm/0004-installation-and-upgrade-workflow.md:
````diff
@@ -0,0 +1,161 @@
+An example execution to deploy a development instance would be:
+
+``` bash
+$ invenio-cli rdm init
````
This should be amended. Due to implementation constraints it has to be:
``` console
invenio-cli --flavour=rdm init
```

The rest is not impacted.

ppanero commented:
Impotant note: No `setup.py` gets generated by the CLI, which means users will not have the option to extend/add entrypoints with custom development. Shall we allow that?

mb-wali commented:
hello guys. i have tried the installation flow for invenio RDM. i dont understand why the invenio cli is out of the project directory. and using pip is not enough to complete the installation flow, its fine when we use the pipenv outside of the project directory but then we will have two pipfiles one inside the project directory and one outside where the invenio-cli is located.
anyway.. it is a bit confusing but. the installation went alright.. also i was able to create the demo records. but when you want to see the specific records for example <https://127.0.0.1:5000/records/5> there is an error:
```
UnboundLocalError
UnboundLocalError: local variable 'pid' referenced before assignment
```
..
thanks guys.. just wanted to give my update on the installation flow... 
Mojib from TUGRAZ.

lnielsen commented:
@mb-wali Thanks a lot for the comments. Much appreciated.

> 1) why the invenio cli is out of the project directory

This will be fixed with the november release. I.e. it will be:
```
invenio-cli init
cd my-rdm
invenio-cli build

I.e. it will work much like ``npm`` and similar tools.

jbenito3 commented on rdm/0004-installation-and-upgrade-workflow.md:
```diff
@@ -0,0 +1,164 @@
+
+Overall design of the installation and upgrade workflows of Invenio RDM.
+
+## Motivation
```
is this thought only for developers? what about sysadmins?
If I am an institution sysadmin and I simply want to install InvenioRDM as it is. How do I do so?
By install, I mean in a server. Not for development purposes. Is this in the scope of this RFC? Just to setup my brain for the rest of it :)

fenekku replied:
The RFC should cover both: a developer that wants to run it locally and a sysadmin (what I often call a 'hoster') that just wants to generate a working repo from the get-go. The main difference will be between main application running in containers behind uwsgi (production) or running on host filesystem (development).

jbenito3 replied:
um...ok, then the motivation list should include some `As a sysadmin...` :)

ppanero commented:
Needs updating:

- `.invenio.private` files.

Possible more.
